### PR TITLE
Tipped arrow color support

### PR DESCRIPTION
**Pull request by heypr**
# Changes
- Modernization of the `ItemColor` class
- Support for tipped arrows added.

Requested by [udnek](https://discord.com/channels/315163488085475337/1164690038307684392) on Discord.

### CODE REVIEW

File: plugin/src/main/java/com/denizenscript/denizen/objects/properties/item/ItemColor.java [hunk 1]
```diff
@@ -3,156 +3,71 @@
...
-                return BukkitColorExtensions.fromColor(mapMeta.getColor()).getObjectAttribute(attribute.fulfill(1));
+            return BukkitColorExtensions.fromColor(mapMeta.getColor());
+        }
+        if (getItemMeta() instanceof PotionMeta potionMeta) {
+            if (!potionMeta.hasColor() || potionMeta.getColor() == null) {
+                return null;
             }
-            return BukkitColorExtensions.fromColor(((LeatherArmorMeta) item.getItemMeta()).getColor()).getObjectAttribute(attribute.fulfill(1));
+            return BukkitColorExtensions.fromColor(potionMeta.getColor());
         }
```
Comment (tal5) [inline plugin/src/main/java/com/denizenscript/denizen/objects/properties/item/ItemColor.java]: Look at the previous handling for `PotionMeta`, need to maintain backwards compatibility

File: plugin/src/main/java/com/denizenscript/denizen/objects/properties/item/ItemColor.java [hunk 2]
```diff
@@ -3,156 +3,71 @@
...
-            if (mat == Material.POTION
-                    || mat == Material.LINGERING_POTION
-                    || mat == Material.SPLASH_POTION) {
-                PotionMeta pm = (PotionMeta) item.getItemMeta();
-                if (!pm.hasColor()) {
-                    return BukkitColorExtensions.fromColor(Color.WHITE).getObjectAttribute(attribute.fulfill((1)));
-                }
-                return BukkitColorExtensions.fromColor(pm.getColor()).getObjectAttribute(attribute.fulfill((1)));
+        if (getItemMeta() instanceof MapMeta mapMeta) {
+            if (!mapMeta.hasColor() || mapMeta.getColor() == null) {
```
Comment (tal5) [inline plugin/src/main/java/com/denizenscript/denizen/objects/properties/item/ItemColor.java]: Any reason you added these `== null` checks instead of just `hasColor`?

File: plugin/src/main/java/com/denizenscript/denizen/objects/properties/item/ItemColor.java [hunk 3]
```diff
@@ -3,156 +3,71 @@
...
-                || mat == Material.SPLASH_POTION
-                || mat == Material.LINGERING_POTION
-                || mat == Material.FILLED_MAP;
-    }
+    // <--[property]
+    // @object ItemTag
+    // @name color
+    // @input ColorTag
+    // @description
+    // Controls the color of the leather armor item, potion item, filled map item, or tipped arrow item.
```
Comment (tal5) [inline plugin/src/main/java/com/denizenscript/denizen/objects/properties/item/ItemColor.java]: Use the property meta format (I.e. what the property actually _is_) - in this case `The color of a leather armor...` or something like that

File: plugin/src/main/java/com/denizenscript/denizen/objects/properties/item/ItemColor.java [hunk 4]
```diff
@@ -3,156 +3,71 @@
...
-            return BukkitColorExtensions.fromColor(mapMeta.getColor()).identify();
+        if (getItemMeta() instanceof MapMeta mapMeta) {
+            mapMeta.setColor(BukkitColorExtensions.getColor(color));
+            getItemStack().setItemMeta(mapMeta);
+            return;
+        }
+        if (getItemMeta() instanceof PotionMeta potionMeta) {
+            potionMeta.setColor(BukkitColorExtensions.getColor(color));
+            getItemStack().setItemMeta(potionMeta);
         }
```
Comment (tal5) [inline plugin/src/main/java/com/denizenscript/denizen/objects/properties/item/ItemColor.java]: Don't really need an `if` for the last check here, as you can know for sure it's that type (since it isn't any of the other ones) - can probably just have a quick in-line `editMeta` call?

File: plugin/src/main/java/com/denizenscript/denizen/objects/properties/item/ItemColor.java [hunk 5]
```diff
@@ -3,156 +3,71 @@
...
-                || mat == Material.SPLASH_POTION) {
-            PotionMeta pm = (PotionMeta) item.getItemMeta();
-            if (!pm.hasColor()) {
-                return null;
-            }
-            return BukkitColorExtensions.fromColor(pm.getColor()).identify();
+    public void setPropertyValue(ColorTag color, Mechanism mechanism) {
+        if (getItemMeta() instanceof LeatherArmorMeta leatherArmorMeta) {
+            leatherArmorMeta.setColor(BukkitColorExtensions.getColor(color));
+            getItemStack().setItemMeta(leatherArmorMeta);
```
Comment (tal5) [inline plugin/src/main/java/com/denizenscript/denizen/objects/properties/item/ItemColor.java]: `ItemProperty` has a `setItemMeta` method, should use that

File: plugin/src/main/java/com/denizenscript/denizen/objects/properties/item/ItemColor.java [hunk 6]
```diff
@@ -3,156 +3,69 @@
...
-                || mat == Material.SPLASH_POTION
-                || mat == Material.LINGERING_POTION
-                || mat == Material.FILLED_MAP;
-    }
+    // <--[property]
+    // @object ItemTag
+    // @name color
+    // @input ColorTag
+    // @description
+    // The color of a leather armor item, potion item, filled map item, or tipped arrow item.
```
Comment (tal5) [inline plugin/src/main/java/com/denizenscript/denizen/objects/properties/item/ItemColor.java]: Nitpick, but would probably be cleaner to have e.g. `of a leather armor, potion, or arrow item` instead of repeating `item` for each?

File: plugin/src/main/java/com/denizenscript/denizen/objects/properties/item/ItemColor.java [hunk 7]
```diff
@@ -3,156 +3,69 @@
...
-                || mat == Material.SPLASH_POTION
-                || mat == Material.LINGERING_POTION
-                || mat == Material.FILLED_MAP;
-    }
+    // <--[property]
+    // @object ItemTag
+    // @name color
+    // @input ColorTag
+    // @description
+    // The color of a leather armor, potion, filled map, or tipped arrow item.
```
Comment (tal5) [inline plugin/src/main/java/com/denizenscript/denizen/objects/properties/item/ItemColor.java:20]: Should mention that it can return null for maps, e.g. `For maps, will return null if the map doesn't have any color.`

File: plugin/src/main/java/com/denizenscript/denizen/objects/properties/item/ItemColor.java [hunk 8]
```diff
@@ -3,156 +3,70 @@
...
-                MapMeta mapMeta = (MapMeta) item.getItemMeta();
-                if (!mapMeta.hasColor()) {
-                    return null;
-                }
-                return BukkitColorExtensions.fromColor(mapMeta.getColor()).getObjectAttribute(attribute.fulfill(1));
+            return BukkitColorExtensions.fromColor(mapMeta.getColor());
+        }
+        if (getItemMeta() instanceof PotionMeta potionMeta) {
+            if (!potionMeta.hasColor()) {
+                return BukkitColorExtensions.fromColor(Color.WHITE);
```
Comment (mcmonkey4eva) [inline plugin/src/main/java/com/denizenscript/denizen/objects/properties/item/ItemColor.java]: Why was this changed?

**Comment by mcmonkey4eva** [comment]
(marked as draft for now) per Discord discussion earlier, need to make the tag and propertyValue impls match their null case handling to the original logic (one was null and the other defaults to white)

**Review by mcmonkey4eva — APPROVED** [review]

**Comment by tal5** [comment]
Reposting from [Discord](https://discord.com/channels/315163488085475337/1011496047811506227/1209224827378868324) - need to use the `PropertyParser` registration methods for the tag/mech, and the mech can just use `setPropertyValue` instead of having duplicate logic

File: plugin/src/main/java/com/denizenscript/denizen/objects/properties/item/ItemColor.java [hunk 9]
```diff
@@ -3,156 +3,93 @@
...
+public class ItemColor extends ItemProperty<ColorTag> {
+
+    // <--[property]
+    // @object ItemTag
+    // @name color
+    // @input ColorTag
+    // @description
+    // The color of a leather armor, potion, filled map, or tipped arrow item.
+    // Will return a white <@link objecttype ColorTag> if the given potion item doesn't have a color.
+    // Will return null if the given map item doesn't have a color.
```
Comment (tal5) [inline plugin/src/main/java/com/denizenscript/denizen/objects/properties/item/ItemColor.java]: Property meta that has mechanism-specific information usually has `For the mechanism:`, should probably have a similar thing for the tag-specific info here (I.e. `For the tag:`)

**Review by tal5 — APPROVED** [review]
